ndb/src/google/cloud/ndb/model.py

```diff
@@ -852,6 +852,146 @@ def _retrieve_value(self, entity, default=None):
         """
         return entity._values.get(self._name, default)
 
+    def _get_user_value(self, entity):
+        """Return the user value for this property of the given entity.
+
+        This implies removing the :class:`_BaseValue` wrapper if present, and
+        if it is, calling all ``_from_base_type()`` methods, in the reverse
+        method resolution order of the property's class. It also handles
+        default values and repeated properties.
+
+        Args:
+            entity (Model): An entity to get a value from.
+
+        Returns:
+            Any: The original value (if not :class:`_BaseValue`) or the wrapped
+            value converted from the base type.
+        """
+        return self._apply_to_values(entity, self._opt_call_from_base_type)
+
+    def _get_base_value(self, entity):
+        """Return the base value for this property of the given entity.
+
+        This implies calling all ``_to_base_type()`` methods, in the method
+        resolution order of the property's class, and adding a
+        :class:`_BaseValue` wrapper, if one is not already present. (If one
+        is present, no work is done.)  It also handles default values and
+        repeated properties.
+
+        Args:
+            entity (Model): An entity to get a value from.
+
+        Returns:
+            Union[_BaseValue, List[_BaseValue]]: The original value
+            (if :class:`_BaseValue`) or the value converted to the base type
+            and wrapped.
+        """
+        return self._apply_to_values(entity, self._opt_call_to_base_type)
+
+    def _get_base_value_unwrapped_as_list(self, entity):
+        """Like _get_base_value(), but always returns a list.
+
+        Args:
+            entity (Model): An entity to get a value from.
+
+        Returns:
+            List[Any]: The unwrapped base values. For an unrepeated
+            property, if the value is missing or :data:`None`, returns
+            ``[None]``; for a repeated property, if the original value is
+            missing or :data:`None` or empty, returns ``[]``.
+
+        Raises:
+            ValueError: If the property is repeated but the base value is not
+                a :class:`list` or :data:`None`.
+            ValueError: If the property is not repeated but the base value is
+                not a :class:`_BaseValue`.
+        """
+        wrapped = self._get_base_value(entity)
+        if self._repeated:
+            if not isinstance(wrapped, list):
+                raise ValueError("Expected the wrapped quantity to be a list.")
```

Comment:
If it were just me, I'd convert back to use an `assert` statement, since that's the best way to express, "Hey, I'm assumiing this will never happen, I want to test my assumption during development, and I want to express to future developers that I at least considered this edge case."

Some orgs hate `assert` statements, though.  I don't know if Google is one of them.

Reply:
I don't know about Google, but I personally am not a fan of `assert` statements. They don't provide a useful error message to users running the code and they can be turned off if the right flag/env. var. is used, which is often a gotcha for devs and users.

Reply:
UPDATE: The Google Python [style guide][1] does not have any issue with `assert` (when used correctly).

[1]: https://github.com/google/styleguide/blob/gh-pages/pyguide.md#244-decision

Reply:
Right. If you think there's any chance the condition would occur IRL, then the code should be left as-is. The `assert` is just a way to document assumptions being made about the code which should always be true, and provide a sanity check during development.  By the time code gets into production, turning them should be OK.  If you still think it could happen IRL, it shouldn't be an assert.  Deleting the check altogether, while fine (I wouldn't argue with you if that's what you eventually did), leaves the assumption implicit and opaque to future developers, whereas an `assert` tells future developers, explicitly, "I'm making this assumption."